alf/algorithms/functional_particle_vi_algorithm_test.py

```diff
@@ -0,0 +1,169 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl.testing import parameterized
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+import alf
+from alf.algorithms.functional_particle_vi_algorithm import FuncParVIAlgorithm
+from alf.tensor_specs import TensorSpec
+from alf.utils import math_ops
+
+
+class FuncParVIAlgorithmTest(parameterized.TestCase, alf.test.TestCase):
+    def cov(self, data, rowvar=False):
+        """Estimate a covariance matrix given data.
+
+        Args:
+            data (tensor): A 1-D or 2-D tensor containing multiple observations 
+                of multiple dimentions. Each row of ``mat`` represents a
+                dimension of the observation, and each column a single
+                observation.
+            rowvar (bool): If True, then each row represents a dimension, with
+                observations in the columns. Othewise, each column represents
+                a dimension while the rows contains observations.
+
+        Returns:
+            The covariance matrix
+        """
+        x = data.detach().clone()
+        if x.dim() > 2:
+            raise ValueError('data has more than 2 dimensions')
+        if x.dim() < 2:
+            x = x.view(1, -1)
+        if not rowvar and x.size(0) != 1:
+            x = x.t()
+        fact = 1.0 / (x.size(1) - 1)
+        x -= torch.mean(x, dim=1, keepdim=True)
+        return fact * x.matmul(x.t()).squeeze()
+
+    def assertArrayGreater(self, x, y, eps):
+        self.assertEqual(x.shape, y.shape)
+        self.assertGreater(float(torch.min(x - y)), eps)
+
+    @parameterized.parameters(('gfsf'), ('svgd'), ('gfsf', True),
+                              ('svgd', True))
+    def test_functional_par_vi_algorithm(self,
+                                         par_vi='svgd',
+                                         function_vi=False,
+                                         num_particles=256,
+                                         train_batch_size=10):
+        """
+        The hypernetwork is trained to generate the parameter vector for a linear
+        regressor. The target linear regressor is :math:`y = X\beta + e`, where 
+        :math:`e\sim N(0, I)` is random noise, :math:`X` is the input data matrix, 
+        and :math:`y` is target ouputs. The posterior of :math:`\beta` has a 
+        closed-form :math:`p(\beta|X,y)\sim N((X^TX)^{-1}X^Ty, X^TX)`.
+        For a linear generator with weight W and bias b, and takes standard Gaussian 
+        noise as input, the output follows a Gaussian :math:`N(b, WW^T)`, which should 
+        match the posterior :math:`p(\beta|X,y)` for both svgd and gfsf.
+        
+        """
+        input_size = 3
+        input_spec = TensorSpec((input_size, ), torch.float32)
+        output_dim = 1
+        batch_size = 100
```

Comment:
batch_size happens to be the square of train_batch_size. This can potentially hide some errors when testing the cov. I suggest you use some other batch_size such as 150

Reply:
Updated.